pkg/grpc/server.go

```diff
@@ -0,0 +1,214 @@
+// Copyright 2017 Envoyproxy Authors
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+// Package server provides an implementation of a streaming xDS server
+package server
+
+import (
+	"strconv"
+	"sync/atomic"
+
+	"github.com/envoyproxy/go-control-plane/api"
+	"github.com/envoyproxy/go-control-plane/pkg/cache"
+	context "golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// Resource types in xDS v2
+const (
+	typePrefix   = "type.googleapis.com/envoy.api.v2."
+	EndpointType = typePrefix + "LbEndpoint"
+	ClusterType  = typePrefix + "Cluster"
+	RouteType    = typePrefix + "Route"
+	ListenerType = typePrefix + "Listener"
+)
+
+// Server is a collection of handlers for streaming discovery requests
+type Server interface {
+	// Register the handlers in a gRPC server
+	Register(*grpc.Server)
+}
+
+// NewServer creates handlers from a config watcher
+func NewServer(config cache.ConfigWatcher) Server {
+	return &server{config: config}
+}
+
+type server struct {
+	config cache.ConfigWatcher
+}
+
+func (s *server) Register(srv *grpc.Server) {
+	api.RegisterAggregatedDiscoveryServiceServer(srv, s)
+	api.RegisterEndpointDiscoveryServiceServer(srv, s)
+	api.RegisterClusterDiscoveryServiceServer(srv, s)
+	api.RegisterRouteDiscoveryServiceServer(srv, s)
+	api.RegisterListenerDiscoveryServiceServer(srv, s)
+}
+
+type stream interface {
+	Send(*api.DiscoveryResponse) error
+	Recv() (*api.DiscoveryRequest, error)
+}
+
+// promises for all xDS resource types
+type promises struct {
+	endpoints cache.Promise
+	clusters  cache.Promise
+	routes    cache.Promise
+	listeners cache.Promise
+}
+
+func (values promises) Cancel() {
```

Comment:
Should values be a pointer to promises?

Reply:
I don't think it makes a big difference here. Promises are intentionally non-pointers, so I went along that line.